### ISSUE TITLE
The difference in loss value between "continue" and "interrupt" training

### ISSUE THREAD

**Issue by svecjan** [speechbrain/speechbrain#2728]
### Describe the bug

I have played with [BEST-RQ recipe](https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech/self-supervised-learning/BEST-RQ), and my goal is replicate results from [article](https://arxiv.org/abs/2405.04296). During my affords, I noticed unusual model behavior through the training process. When I kept training "continuously" (`epoch1 -> epoch2 -> epoch3 -> ...`), the model had a worse performance from the point of loss value than when the model was trained "interruptedly" (`epoch1 -> [interrapt, start training from epoch1's checkpoint] -> epoch2 -> [interrapt, start training from epoch2's checkpoint] -> epoch3 -> ...`).

### Expected behaviour

Loss value progress stays invariant in the training process when the starting checkpoint is similar.

### To Reproduce

- Run as:
```
python train.py hparams/BEST-RQ.yaml --find_unused_parameters --data_folder /path/to/LibriSpeech --output_folder exp/brq_debug --seconds_per_batch 200 --grad_accumulation_factor 2 --precision fp32 --max_grad_norm 1 --norm_type global
```
- The setup is single node(machine) with 8GPUs 16G.

### Environment Details

```
SpeechBrain system description
==============================
Python version:
3.11.10 | packaged by conda-forge | (main, Sep 22 2024, 14:10:38) [GCC 13.3.0]
==============================
Installed Python packages:
asttokens==2.4.1
black==24.3.0
certifi==2024.8.30
cfgv==3.4.0
charset-normalizer==3.3.2
click==8.1.7
decorator==5.1.1
distlib==0.3.8
docstring_parser_fork==0.0.9
executing==2.1.0
filelock==3.16.1
flake8==7.0.0
fsspec==2024.9.0
huggingface-hub==0.25.1
HyperPyYAML==1.2.2
identify==2.6.1
idna==3.10
iniconfig==2.0.0
ipdb==0.13.13
ipython==8.28.0
isort==5.13.2
jedi==0.19.1
Jinja2==3.1.4
joblib==1.4.2
llvmlite==0.43.0
MarkupSafe==2.1.5
matplotlib-inline==0.1.7
mccabe==0.7.0
mpmath==1.3.0
mypy-extensions==1.0.0
networkx==3.3
nodeenv==1.9.1
numba @ file:///home/conda/feedstock_root/build_artifacts/numba_1718888020163/work
numpy==1.26.4
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu12==12.1.0.106
nvidia-nccl-cu12==2.20.5
nvidia-nvjitlink-cu12==12.6.68
nvidia-nvtx-cu12==12.1.105
packaging==24.1
pandas==2.2.3
parso==0.8.4
pathspec==0.12.1
pexpect==4.9.0
platformdirs==4.3.6
pluggy==1.5.0
pre-commit==3.8.0
prompt_toolkit==3.0.48
ptyprocess==0.7.0
pure_eval==0.2.3
pycodestyle==2.11.0
pydoclint==0.4.1
pyflakes==3.2.0
Pygments==2.18.0
pygtrie==2.5.0
pytest==7.4.0
python-dateutil==2.9.0.post0
pytz==2024.2
PyYAML==6.0.2
regex==2024.9.11
requests==2.32.3
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
safetensors==0.4.5
scipy==1.12.0
sentencepiece==0.2.0
six==1.16.0
-e git+https://github.com/speechbrain/speechbrain.git@1bb368a016aac2b7d3cc544622d44ee5bf4897be#egg=speechbrain
stack-data==0.6.3
sympy==1.13.3
tokenizers==0.19.1
torch==2.4.1
torchaudio==2.4.1
tqdm==4.66.5
traitlets==5.14.3
transformers==4.44.2
triton==3.0.0
typing_extensions==4.12.2
tzdata==2024.2
urllib3==2.2.3
virtualenv==20.26.5
wcwidth==0.2.13
yamllint==1.35.1
==============================
Git revision:
1bb368a01
==============================
CUDA version:
12.1
```

### Relevant Log Output

```shell
Logs are below.
```

### Additional Context

- "Continue" training
```
epoch: 1, step: 2265, lr: 3.62e-05, loss: 6.233237266540527, avg_train_loss_old: 7.18, avg_train_loss_new: 7.18 wavs: torch.Size([12, 257761])
epoch: 1, step: 2266, lr: 3.63e-05, loss: 6.438640117645264, avg_train_loss_old: 7.18, avg_train_loss_new: 7.18 wavs: torch.Size([13, 240960])
epoch: 2, step: 1, lr: 3.63e-05, loss: 6.361218452453613, avg_train_loss_old: 0.00e+00, avg_train_loss_new: 6.36 wavs: torch.Size([15, 200080])
epoch: 2, step: 2, lr: 3.63e-05, loss: 6.379264831542969, avg_train_loss_old: 6.36, avg_train_loss_new: 6.37 wavs: torch.Size([12, 249280])
epoch: 2, step: 3, lr: 3.63e-05, loss: 6.2642974853515625, avg_train_loss_old: 6.37, avg_train_loss_new: 6.33 wavs: torch.Size([13, 237120])
epoch: 2, step: 4, lr: 3.63e-05, loss: 6.255147933959961, avg_train_loss_old: 6.33, avg_train_loss_new: 6.31 wavs: torch.Size([18, 176800])
epoch: 2, step: 5, lr: 3.63e-05, loss: 6.301097869873047, avg_train_loss_old: 6.31, avg_train_loss_new: 6.31 wavs: torch.Size([12, 253120])
epoch: 2, step: 6, lr: 3.64e-05, loss: 6.311615943908691, avg_train_loss_old: 6.31, avg_train_loss_new: 6.31 wavs: torch.Size([35, 90800])
epoch: 2, step: 7, lr: 3.64e-05, loss: 6.237408638000488, avg_train_loss_old: 6.31, avg_train_loss_new: 6.30 wavs: torch.Size([14, 225120])
epoch: 2, step: 8, lr: 3.64e-05, loss: 6.136245250701904, avg_train_loss_old: 6.30, avg_train_loss_new: 6.28 wavs: torch.Size([13, 241200])
epoch: 2, step: 9, lr: 3.64e-05, loss: 6.482987403869629, avg_train_loss_old: 6.28, avg_train_loss_new: 6.30 wavs: torch.Size([20, 158160])
epoch: 2, step: 10, lr: 3.64e-05, loss: 6.327013969421387, avg_train_loss_old: 6.30, avg_train_loss_new: 6.31 wavs: torch.Size([11, 270800])
```
- "Interrupt" training
```
epoch: 1, step: 2265, lr: 3.62e-05, loss: 6.266966819763184, avg_train_loss_old: 7.19, avg_train_loss_new: 7.19 wavs: torch.Size([12, 257761])
epoch: 1, step: 2266, lr: 3.63e-05, loss: 6.451587200164795, avg_train_loss_old: 7.19, avg_train_loss_new: 7.19 wavs: torch.Size([13, 240960])
epoch: 2, step: 1, lr: 3.63e-05, loss: 6.383120536804199, avg_train_loss_old: 0.00e+00, avg_train_loss_new: 6.38 wavs: torch.Size([15, 200080])
epoch: 2, step: 2, lr: 3.63e-05, loss: 6.337033271789551, avg_train_loss_old: 6.38, avg_train_loss_new: 6.36 wavs: torch.Size([12, 249280])
epoch: 2, step: 3, lr: 3.63e-05, loss: 6.130838394165039, avg_train_loss_old: 6.36, avg_train_loss_new: 6.28 wavs: torch.Size([13, 237120])
epoch: 2, step: 4, lr: 3.63e-05, loss: 6.232778549194336, avg_train_loss_old: 6.28, avg_train_loss_new: 6.27 wavs: torch.Size([18, 176800])
epoch: 2, step: 5, lr: 3.63e-05, loss: 6.06861686706543, avg_train_loss_old: 6.27, avg_train_loss_new: 6.23 wavs: torch.Size([12, 253120])
epoch: 2, step: 6, lr: 3.64e-05, loss: 6.164124965667725, avg_train_loss_old: 6.23, avg_train_loss_new: 6.22 wavs: torch.Size([35, 90800])
epoch: 2, step: 7, lr: 3.64e-05, loss: 5.932363510131836, avg_train_loss_old: 6.22, avg_train_loss_new: 6.18 wavs: torch.Size([14, 225120])
epoch: 2, step: 8, lr: 3.64e-05, loss: 5.954776763916016, avg_train_loss_old: 6.18, avg_train_loss_new: 6.15 wavs: torch.Size([13, 241200])
epoch: 2, step: 9, lr: 3.64e-05, loss: 6.263972759246826, avg_train_loss_old: 6.15, avg_train_loss_new: 6.16 wavs: torch.Size([20, 158160])
epoch: 2, step: 10, lr: 3.64e-05, loss: 5.755927085876465, avg_train_loss_old: 6.16, avg_train_loss_new: 6.12 wavs: torch.Size([11, 270800])
```
... and jump on end of 2th and start 3rd epoch
- "Continue" training
```
epoch: 2, step: 2264, lr: 7.25e-05, loss: 5.856682300567627, avg_train_loss_old: 6.01, avg_train_loss_new: 6.01 wavs: torch.Size([34, 93520])
epoch: 2, step: 2265, lr: 7.25e-05, loss: 5.924241065979004, avg_train_loss_old: 6.01, avg_train_loss_new: 6.01 wavs: torch.Size([29, 107040])
epoch: 2, step: 2266, lr: 7.25e-05, loss: 5.866128921508789, avg_train_loss_old: 6.01, avg_train_loss_new: 6.01 wavs: torch.Size([13, 236720])
epoch: 3, step: 1, lr: 7.25e-05, loss: 5.927073001861572, avg_train_loss_old: 0.00e+00, avg_train_loss_new: 5.93 wavs: torch.Size([14, 217760])
epoch: 3, step: 2, lr: 7.25e-05, loss: 5.7725701332092285, avg_train_loss_old: 5.93, avg_train_loss_new: 5.85 wavs: torch.Size([29, 109680])
epoch: 3, step: 3, lr: 7.25e-05, loss: 5.7523651123046875, avg_train_loss_old: 5.85, avg_train_loss_new: 5.82 wavs: torch.Size([13, 229280])
```
- "Interrupt" training
```
epoch: 2, step: 2264, lr: 7.25e-05, loss: 3.595381021499634, avg_train_loss_old: 3.92, avg_train_loss_new: 3.92 wavs: torch.Size([34, 93520])
epoch: 2, step: 2265, lr: 7.25e-05, loss: 3.6582729816436768, avg_train_loss_old: 3.92, avg_train_loss_new: 3.92 wavs: torch.Size([29, 107040])
epoch: 2, step: 2266, lr: 7.25e-05, loss: 3.6624903678894043, avg_train_loss_old: 3.92, avg_train_loss_new: 3.92 wavs: torch.Size([13, 236720])
epoch: 3, step: 1, lr: 7.25e-05, loss: 3.635432720184326, avg_train_loss_old: 0.00e+00, avg_train_loss_new: 3.64 wavs: torch.Size([14, 217760])
epoch: 3, step: 2, lr: 7.25e-05, loss: 3.6585352420806885, avg_train_loss_old: 3.64, avg_train_loss_new: 3.65 wavs: torch.Size([29, 109680])
epoch: 3, step: 3, lr: 7.25e-05, loss: 3.418518304824829, avg_train_loss_old: 3.65, avg_train_loss_new: 3.57 wavs: torch.Size([13, 229280])
```
- Description:
  - Values `loss, avg_train_loss_old, avg_train_loss_new` are parameters/return of `speechbrain.core.Brain.update_average` function as `avg_train_loss_new = update_average(loss, avg_train_loss_old)
  - Value `wavs` is tensor shape from `speechbrain.core.Brain.compute_forward` (showing sequence of batch is same).
- Last words:
  - I didn't try another recipe, so I don't know if it affects them too.
  - I would like to try running with only one GPU to rule out synchronization between GPUs.

**Comment by SevKod** [comment]
I am also playing with the recipe, and I experienced something similar during the training of the entire model. 

I trained BEST-RQ using 8 a40-48gb for a first 30 hours. The validation loss stagnated at 5.80 (and validation accuacy as well, at around 3%). Then I resumed training for another 30 hours and all my losses spiked down as you can see below : 

> epoch: 125, steps: 70125, lr: 4.78e-04 - train loss: 5.81 - valid loss: 6.36, valid accuracy: 2.79e-02
> epoch: 126, steps: 70686, lr: 4.76e-04 - train loss: 5.81 - valid loss: 6.37, valid accuracy: 2.67e-02
> epoch: 127, steps: 71247, lr: 4.74e-04 - train loss: 5.81 - valid loss: 6.43, valid accuracy: 2.43e-02
> epoch: 128, steps: 71808, lr: 4.72e-04 - train loss: 5.81 - valid loss: 6.38, valid accuracy: 2.54e-02
> epoch: 129, steps: 72369, lr: 4.70e-04 - train loss: 5.81 - valid loss: 6.37, valid accuracy: 3.27e-02
> epoch: 130, steps: 72930, lr: 4.68e-04 - train loss: 5.81 - valid loss: 6.31, valid accuracy: 3.94e-02
> epoch: 131, steps: 73491, lr: 4.67e-04 - train loss: 5.80 - valid loss: 6.34, valid accuracy: 2.53e-02
> epoch: 132, steps: 74052, lr: 4.65e-04 - train loss: 5.81 - valid loss: 6.38, valid accuracy: 3.10e-02
> epoch: 133, steps: 74613, lr: 4.63e-04 - train loss: 3.83 - valid loss: 4.32, valid accuracy: 1.71e-01
> epoch: 134, steps: 75174, lr: 4.61e-04 - train loss: 3.73 - valid loss: 4.24, valid accuracy: 1.72e-01
> epoch: 135, steps: 75735, lr: 4.60e-04 - train loss: 3.70 - valid loss: 4.24, valid accuracy: 1.72e-01
> epoch: 136, steps: 76296, lr: 4.58e-04 - train loss: 3.70 - valid loss: 4.22, valid accuracy: 1.76e-01
> epoch: 137, steps: 76857, lr: 4.56e-04 - train loss: 3.69 - valid loss: 4.18, valid accuracy: 1.75e-01
> epoch: 138, steps: 77418, lr: 4.55e-04 - train loss: 3.68 - valid loss: 4.20, valid accuracy: 1.75e-01
> epoch: 139, steps: 77979, lr: 4.53e-04 - train loss: 3.67 - valid loss: 4.23, valid accuracy: 1.72e-01

Accuracy went up from 3% to 17%

I did not experience this on a single gpu though, so it might be linked to gpu synchronization. This behavior only happened once, and did not happen again after resuming the training again. In fact, the validation loss did not go below 4.14 (even after 300+ epochs), and the max accuracy I could get was 18%.

**Comment by whettenr** [comment]
@svecjan @SevKod @TParcollet 
Hi everyone,

Sorry it has taken me so long to get to this. But I was able to reproduce and find the issue, and turns out this actually solves another big issue. We are working on finding the best solution ASAP.

Sorry again,
Ryan

**Comment by whettenr** [comment]
Hi everyone @svecjan @SevKod 
With a recent merge from @TParcollet (#2814), this issue should be fixed! 

Thanks for pointing this out!

Ryan
